pkg/networkserver/downlink.go

```diff
@@ -171,35 +171,42 @@ func (ns *NetworkServer) generateDataDownlink(ctx context.Context, dev *ttnpb.En
 	}
 	maxDownLen, maxUpLen = maxDownLen-8, maxUpLen-8
 
-	var fPending bool
-	spec := lorawan.DefaultMACCommands
-	cmds := make([]*ttnpb.MACCommand, 0, len(dev.MACState.QueuedResponses)+len(dev.MACState.PendingRequests))
-	var lostResps []*ttnpb.MACCommand
+	var (
+		fPending bool
+		genState generateDownlinkState
+		cmdBuf   []byte
+	)
 	if class == ttnpb.CLASS_A {
-		for i, cmd := range dev.MACState.QueuedResponses {
-			desc := spec[cmd.CID]
-			if desc == nil {
-				lostResps = append(lostResps, dev.MACState.QueuedResponses[i:]...)
+		spec := lorawan.DefaultMACCommands
+		cmds := make([]*ttnpb.MACCommand, 0, len(dev.MACState.QueuedResponses)+len(dev.MACState.PendingRequests))
+
+		for _, cmd := range dev.MACState.QueuedResponses {
+			logger := logger.WithField("cid", cmd.CID)
+			desc, ok := spec[cmd.CID]
+			switch {
+			case !ok:
+				logger.Error("Unknown MAC command response enqueued, set FPending")
 				maxDownLen = 0
 				fPending = true
-				break
-			}
-			if desc.DownlinkLength > maxDownLen {
```

Comment:
Note the bug fixed here - `desc.DownlinkLength` only contains the command payload length, but we should also account for the CID